Added Infernal Painting

caelwarner opened the pull request:
- Infernal Painting will only show paintings from IE
- Vanilla paintings won't show IE paintings
- Infernal paintings still need custom texture, recipe and lang file line

caelwarner committed:
Added Infernal Painting

- Infernal Painting will only show paintings from IE
- Vanilla paintings won't show IE paintings
- Infernal paintings still need custom texture, recipe and lang file line

caelwarner committed:
Merge branch 'master-indev' into infernal-painting

# Conflicts:
#	src/main/java/com/nekomaster1000/infernalexp/init/IEItems.java
#	src/main/resources/infernal-expansion.mixins.json

caelwarner committed:
Added custom item texture, back texture, recipe and lang file entry for infernal painting

- Added InfernalPaintingRenderer to render custom back texture
- Added NetworkManager to handle packets (this will need to be changed to the system that Hellion created in #42)
- Added SpawnInfernalPaintingPacket to spawn an InfernalPaintingEntity instead of PaintingEntity
- Added MixinPaintingSpriteUploader to add custom back texture to painting sprite atlas

caelwarner committed:
Fixed issues with dedicated server

- Stopped dedicated server from accessing ClientWorld

caelwarner committed:
Merge branch 'master-indev' into infernal-painting

# Conflicts:
#	src/main/java/com/nekomaster1000/infernalexp/InfernalExpansion.java
#	src/main/java/com/nekomaster1000/infernalexp/init/IEItems.java
#	src/main/resources/META-INF/accesstransformer.cfg
#	src/main/resources/assets/infernalexp/lang/en_us.json
#	src/main/resources/infernal-expansion.mixins.json

caelwarner committed:
Infernal Paintings now use the new packet system (Ready for merge)

- Switched over to our new network packet system

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/InfernalPaintingEntity.java:
```diff
@@ -0,0 +1,100 @@
+		this.hangingPosition = pos;
+		updateFacingWithBoundingBox(facing);
+
+		List<PaintingType> paintings = new ArrayList<>();
```
Wouldn't a `Set` be more appropriate?

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/entities/InfernalPaintingEntity.java:
```diff
@@ -0,0 +1,100 @@
+		this.hangingPosition = pos;
+		updateFacingWithBoundingBox(facing);
+
+		List<PaintingType> paintings = new ArrayList<>();
```
Wouldn't a Set be more appropriate?

caelwarner replied:
Its much easier to get a random item from a list than it is from a set

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/init/IEEntityTypes.java:
```diff
@@ -77,24 +77,29 @@
-    public static void register(IEventBus eventBus) {
-        ENTITY_TYPES.register(eventBus);
-        InfernalExpansion.LOGGER.info("Infernal Expansion: Entity Types Registered!");
+	public static final RegistryObject<EntityType<AscusBombEntity>> ASCUS_BOMB = ENTITY_TYPES.register("ascus_bomb",
+		() -> EntityType.Builder.<AscusBombEntity>create(AscusBombEntity::new, EntityClassification.MISC)
+			.size(0.25f, 0.25f).trackingRange(4).updateInterval(10)
+			.build(new ResourceLocation(InfernalExpansion.MOD_ID, "ascus_bomb").toString()));
+
+	public static final RegistryObject<EntityType<ThrowableMagmaCreamEntity>> THROWABLE_MAGMA_CREAM = ENTITY_TYPES.register("throwable_magma_cream",
+		() -> EntityType.Builder.<ThrowableMagmaCreamEntity>create(ThrowableMagmaCreamEntity::new, EntityClassification.MISC)
+			.size(0.25f, 0.25f).trackingRange(4).updateInterval(10)
+			.build(new ResourceLocation(InfernalExpansion.MOD_ID, "throwable_magma_cream").toString()));
```
stop indenting stuff that isn't related to the pr

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/init/IEEntityTypes.java:
```diff
@@ -77,24 +77,29 @@
+			.size(0.5F, 0.5F).trackingRange(10).updateInterval(Integer.MAX_VALUE)
+			.build(new ResourceLocation(InfernalExpansion.MOD_ID, "infernal_painting").toString()));
+
+	private static <T extends Entity> RegistryObject<EntityType<T>> register(String key, EntityType.IFactory<T> factoryIn, EntityClassification classificationIn, float size1, float size2) {
+		return ENTITY_TYPES.register(key, () -> EntityType.Builder.create(factoryIn, classificationIn)
+			.size(size1, size2)
+			.build(new ResourceLocation(InfernalExpansion.MOD_ID, key).toString()));
+	}
+
+	public static void register(IEventBus eventBus) {
+		ENTITY_TYPES.register(eventBus);
+		InfernalExpansion.LOGGER.info("Infernal Expansion: Entity Types Registered!");
```
Same as comment on lines +80 to +88

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/init/IEItems.java:
```diff
@@ -208,13 +210,14 @@
 
-	public static final RegistryObject<Item> SHROOMLIGHT_FUNGUS = ITEMS.register("shroomlight_fungus",      () -> new BlockItemBase(IEBlocks.SHROOMLIGHT_FUNGUS.get()));
-	public static final RegistryObject<Item> LUMINOUS_FUNGUS = ITEMS.register("luminous_fungus",    () -> new BlockItemBase(IEBlocks.LUMINOUS_FUNGUS.get()));
+	public static final RegistryObject<Item> SHROOMLIGHT_FUNGUS = ITEMS.register("shroomlight_fungus", () -> new BlockItemBase(IEBlocks.SHROOMLIGHT_FUNGUS.get()));
+	public static final RegistryObject<Item> LUMINOUS_FUNGUS = ITEMS.register("luminous_fungus", () -> new BlockItemBase(IEBlocks.LUMINOUS_FUNGUS.get()));
 
-	public static final RegistryObject<Item> DULLTHORNS = ITEMS.register("dullthorns",		      () -> new DullthornsBlockItem(IEBlocks.DULLTHORNS.get()));
-	public static final RegistryObject<Item> DULLTHORNS_BLOCK = ITEMS.register("dullthorns_block",  () -> new BlockItemBase(IEBlocks.DULLTHORNS_BLOCK.get()));
+	public static final RegistryObject<Item> DULLTHORNS = ITEMS.register("dullthorns", () -> new DullthornsBlockItem(IEBlocks.DULLTHORNS.get()));
+	public static final RegistryObject<Item> DULLTHORNS_BLOCK = ITEMS.register("dullthorns_block", () -> new BlockItemBase(IEBlocks.DULLTHORNS_BLOCK.get()));
```
stop indenting stuff that is not in your pr

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/init/IEPaintings.java:
```diff
@@ -1,23 +1,22 @@
+	public static RegistryObject<PaintingType> THE_FALLEN_ONES = PAINTING_TYPES.register("the_fallen_ones", () -> new PaintingType(48, 64));
+	public static RegistryObject<PaintingType> CHILLING_ISLES = PAINTING_TYPES.register("chilling_isles", () -> new PaintingType(48, 48));
+	public static RegistryObject<PaintingType> VOLINE = PAINTING_TYPES.register("voline", () -> new PaintingType(16, 16));
+	public static RegistryObject<PaintingType> DA_SALT = PAINTING_TYPES.register("da_salt", () -> new PaintingType(64, 32));
```
yeah yeah
![image](https://user-images.githubusercontent.com/36593309/113840210-20bd4f00-9799-11eb-937e-d458259ed256.png)

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/items/InfernalPaintingItem.java:
```diff
@@ -0,0 +1,44 @@
+
+			CompoundNBT compoundNBT = context.getItem().getTag();
+			if (compoundNBT != null) {
+				EntityType.applyItemNBT(context.getWorld(), context.getPlayer(), paintingEntity, compoundNBT);
```
possible NullPointerException at `context.getPlayer()`, since you didn't check if it isn't null

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/items/InfernalPaintingItem.java:
```diff
@@ -0,0 +1,44 @@
+	public ActionResultType onItemUse(ItemUseContext context) {
+		BlockPos blockPos = context.getPos().offset(context.getFace());
+
+		if (context.getPlayer() != null && !this.canPlace(context.getPlayer(), context.getFace(), context.getItem(), blockPos)) {
```
```suggestion
		if (context.getPlayer() == null || !this.canPlace(context.getPlayer(), context.getFace(), context.getItem(), blockPos)) {
```
resolves my comment about a NullPointerException on line 28

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/mixin/common/MixinPaintingEntity.java:
```diff
@@ -0,0 +1,35 @@
+	// Minecraft always wants to place the biggest painting possible, this makes sure that it doesn't think that an Infernal Expansion painting is the biggest possible
+	@ModifyVariable(at = @At(value = "STORE"), method = "<init>(Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/util/Direction;)V", name = "j", ordinal = 1)
+	public int IE_changeMaxSurfaceArea(int maxSurfaceArea) {
+		if (art.getRegistryName().getNamespace().equals("infernalexp")) {
+			return 0;
+		}
+
+		return maxSurfaceArea;
```
```suggestion
		return art.getRegistryName().getNamespace().equals("infernalexp") ? 0 : maxSurfaceArea;
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/network/IENetworkHandler.java:
```diff
@@ -19,8 +20,9 @@
 
 	public static synchronized void register() {
-		INSTANCE.messageBuilder(WhipReachPacket.class, index++).encoder(WhipReachPacket::encode).decoder(WhipReachPacket::decode).consumer(WhipReachPacket::handle).add();
-	}
+        INSTANCE.messageBuilder(WhipReachPacket.class, index++).encoder(WhipReachPacket::encode).decoder(WhipReachPacket::decode).consumer(WhipReachPacket::handle).add();
```
stop indenting shit that isn't in your pr

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/network/IENetworkHandler.java:
```diff
@@ -8,6 +8,7 @@
 import net.minecraftforge.fml.network.simple.SimpleChannel;
 
 public class IENetworkHandler {
+
```
```suggestion
```
why

caelwarner replied:
it looks nicer

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/util/ClientEventBusSubscriber.java:
```diff
@@ -39,25 +40,26 @@ public static void onClientSetup(FMLClientSetupEvent event) {
-        RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.GLOWSILK_MOTH.get(), GlowsilkMothRenderer::new);
-        RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.ASCUS_BOMB.get(), manager -> new SpriteRenderer<>(manager, Minecraft.getInstance().getItemRenderer()));
-        RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.THROWABLE_MAGMA_CREAM.get(), manager -> new SpriteRenderer<>(manager, Minecraft.getInstance().getItemRenderer()));
+		RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.EMBODY.get(), EmbodyRenderer::new);
+		RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.BASALT_GIANT.get(), BasaltGiantRenderer::new);
+		RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.BLACKSTONE_DWARF.get(), BlackstoneDwarfRenderer::new);
+		RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.GLOWSQUITO.get(), GlowsquitoRenderer::new);
+		RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.SKELETAL_PIGLIN.get(), SkeletalPiglinRenderer::new);
+		RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.PYRNO.get(), PyrnoRenderer::new);
+		RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.BLINDSIGHT.get(), BlindsightRenderer::new);
+		RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.GLOWSILK_MOTH.get(), GlowsilkMothRenderer::new);
+		RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.ASCUS_BOMB.get(), manager -> new SpriteRenderer<>(manager, Minecraft.getInstance().getItemRenderer()));
+		RenderingRegistry.registerEntityRenderingHandler(IEEntityTypes.THROWABLE_MAGMA_CREAM.get(), manager -> new SpriteRenderer<>(manager, Minecraft.getInstance().getItemRenderer()));
```
i stg i will start throwing shit

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/util/ClientEventBusSubscriber.java:
```diff
@@ -39,25 +40,26 @@ public static void onClientSetup(FMLClientSetupEvent event) {
+		ClientRegistry.bindTileEntityRenderer(IETileEntityTypes.GLOW_CAMPFIRE.get(), GlowCampfireTileEntityRenderer::new);
 
-        RenderTypeLookup.setRenderLayer(IEBlocks.LUMINOUS_FUNGUS.get(), RenderType.getCutout());
-        RenderTypeLookup.setRenderLayer(IEBlocks.DULLTHORNS.get(), RenderType.getCutout());
+		RenderTypeLookup.setRenderLayer(IEBlocks.LUMINOUS_FUNGUS.get(), RenderType.getCutout());
+		RenderTypeLookup.setRenderLayer(IEBlocks.DULLTHORNS.get(), RenderType.getCutout());
 
-        RenderTypeLookup.setRenderLayer(IEBlocks.POTTED_LUMINOUS_FUNGUS.get(), RenderType.getCutout());
-        RenderTypeLookup.setRenderLayer(IEBlocks.POTTED_DULLTHORNS.get(), RenderType.getCutout());
-        RenderTypeLookup.setRenderLayer(IEBlocks.POTTED_SHROOMLIGHT_FUNGUS.get(), RenderType.getCutout());
+		RenderTypeLookup.setRenderLayer(IEBlocks.POTTED_LUMINOUS_FUNGUS.get(), RenderType.getCutout());
+		RenderTypeLookup.setRenderLayer(IEBlocks.POTTED_DULLTHORNS.get(), RenderType.getCutout());
+		RenderTypeLookup.setRenderLayer(IEBlocks.POTTED_SHROOMLIGHT_FUNGUS.get(), RenderType.getCutout());
```
i will start smashing shit, including your mom

caelwarner replied:
lines like these were edited before the .editorconfig change to spaces instead of tabs

SwanX1 commented on src/main/resources/META-INF/accesstransformer.cfg:
```diff
@@ -8,5 +8,7 @@ public-f net.minecraft.client.AbstractOption func_243220_a()Lnet/minecraft/util/
 public-f net.minecraft.client.AbstractOption func_243222_a(Lnet/minecraft/util/text/ITextComponent;)Lnet/minecraft/util/text/ITextComponent; # getGenericValueComponent
 public-f net.minecraft.client.gui.widget.TextFieldWidget func_212951_d(Ljava/lang/String;)V # onTextChanged
+public-f net.minecraft.client.renderer.texture.SpriteUploader func_215282_a(Lnet/minecraft/util/ResourceLocation;)Lnet/minecraft/client/renderer/texture/TextureAtlasSprite; # getSprite (SpriteUploader)
+protected-f net.minecraft.client.renderer.entity.PaintingRenderer func_229122_a_(Lcom/mojang/blaze3d/matrix/MatrixStack;Lcom/mojang/blaze3d/vertex/IVertexBuilder;Lnet/minecraft/entity/item/PaintingEntity;IILnet/minecraft/client/renderer/texture/TextureAtlasSprite;Lnet/minecraft/client/renderer/texture/TextureAtlasSprite;)V
```
where comment

caelwarner replied:
the function has no mapping name, its just called func_229122_a_

SwanX1 commented on src/main/resources/assets/infernalexp/lang/en_us.json:
```diff
@@ -155,6 +155,10 @@
   "item.infernalexp.glowsilk": "Glowsilk",
   "item.infernalexp.soul_salt_clump": "Soul Salt Clump",
+  "item.infernalexp.infernal_painting": "Infernal Painting",
+  "item.infernalexp.strider_bucket": "Bucket of Strider",
+  "item.infernalexp.magma_cube_bucket": "Bucket of Magma Cube",
+  "item.infernalexp.voline_bucket": "Bucket of Voline",
```
```suggestion
```
not for this pr

caelwarner replied:
ok

caelwarner committed:
Applied some of the actual suggestions

- All of the weird spacing stuff was automatically change by my IDE before the .editorconfig changes

Co-authored-by: SwanX1 <karliscern@gmail.com>

SwanX1 commented:
LGTM, merge at your own discretion.
